File: jcms/jcms-core/src/main/java/com/ysh/jcms/svc/log/CmsLogStatusValue.java
package com.ysh.jcms.svc.log;

import com.ysh.jcms.core.CmsTypeOld;
import com.ysh.jcms.data.common.CmsEntryId;
import com.ysh.jcms.data.time.CmsBinaryTime;
import java.util.Arrays;
import java.util.List;

/**
 * LogStatusValue ::= SEQUENCE { oldEntrTm [0] IMPLICIT EntryTime, newEntrTm [1]
 * IMPLICIT EntryTime, oldEntr [2] IMPLICIT EntryID, newEntr [3] IMPLICIT
 * EntryID } — 8.8.6
 *
 * Used by GetLogStatusValues response.
 */
public class CmsLogStatusValue extends CmsTypeOld {

    public CmsBinaryTime oldEntrTm;
    public CmsBinaryTime newEntrTm;
    public CmsEntryId oldEntr;
    public CmsEntryId newEntr;

    public CmsLogStatusValue() {
        this.oldEntrTm = new CmsBinaryTime();
        this.newEntrTm = new CmsBinaryTime();
        this.oldEntr = new CmsEntryId();
        this.newEntr = new CmsEntryId();
    }

    public CmsLogStatusValue oldEntrTm(CmsBinaryTime v) {
        this.oldEntrTm = v;
        return this;
    }
    public CmsLogStatusValue newEntrTm(CmsBinaryTime v) {
        this.newEntrTm = v;
        return this;
    }
    public CmsLogStatusValue oldEntr(byte[] v) {
        this.oldEntr.value(v);
        return this;
    }
    public CmsLogStatusValue oldEntr(String v) {
        this.oldEntr.value(v);
        return this;
    }
    public CmsLogStatusValue newEntr(byte[] v) {
        this.newEntr.value(v);
        return this;
    }
    public CmsLogStatusValue newEntr(String v) {
        this.newEntr.value(v);
        return this;
    }

    @Override
    public List<? extends CmsTypeOld> children() {
        return Arrays.asList(oldEntrTm, newEntrTm, oldEntr, newEntr);
    }
}

File: ccms/src/svc/log/cms_log_status_value.c
#include "svc/log/cms_log_status_value.h"
#include "data/common/cms_entry_id.h"
#include "data/common/cms_entry_time.h"

int cms_log_status_value_encode_stream(per_stream_t *s, const cms_log_status_value_t *v) {
    if (!v || !v->old_entr_tm || !v->new_entr_tm || !v->old_entr || !v->new_entr)
        return CMS_ERR;
    int err;

    err = cms_entry_time_encode_stream(s, v->old_entr_tm);
    if (err)
        return err;

    err = cms_entry_time_encode_stream(s, v->new_entr_tm);
    if (err)
        return err;

    err = cms_entry_id_encode_stream(s, v->old_entr);
    if (err)
        return err;

    err = cms_entry_id_encode_stream(s, v->new_entr);
    if (err)
        return err;

    return CMS_OK;
}

int cms_log_status_value_decode_stream(per_stream_t *s, void *ptr) {
    cms_log_status_value_t *v = (cms_log_status_value_t *) ptr;
    int err;

    if (v && (!v->old_entr_tm || !v->new_entr_tm || !v->old_entr || !v->new_entr))
        return CMS_ERR;
    err = cms_entry_time_decode_stream(s, v ? v->old_entr_tm : NULL);
    if (err)
        return err;

    err = cms_entry_time_decode_stream(s, v ? v->new_entr_tm : NULL);
    if (err)
        return err;

    err = cms_entry_id_decode_stream(s, v ? v->old_entr : NULL);
    if (err)
        return err;

    err = cms_entry_id_decode_stream(s, v ? v->new_entr : NULL);
    if (err)
        return err;

    return CMS_OK;
}
